### ISSUE TITLE
Revapi is pretty slow on big jars

### ISSUE THREAD

**Issue by rsvoboda** [revapi/revapi#124]
Revapi is pretty slow on big jars

I did experiment on all-in-one jar for WildFly and it took 25 minutes to get the report
The same bits were processed with japicmp in 1 minute and 12 seconds.
https://github.com/rsvoboda/rsvoboda-playground/tree/master/revapi-and-wildfly#generate-api-diff

**Comment by metlos** [comment]
Is that with Java 9 or Java 8?

**Comment by rsvoboda** [comment]
I think I tried Java 8, I can double check.

**Comment by metlos** [comment]
Well, in either case, I think revapi is always going to be slower than japicmp because it does much more thorough analysis of the APIs - especially computing the API "surface" by use-chain analysis is quite expensive. Revapi understands the distinction between "primary" API and dependencies. It tries to find whether the primary API doesn't expose parts of the dependencies in its own API. This is something japicmp doesn't have even conceptually. Thanks to this use-chain analysis, Revapi is also able to find use-chains through which non-public classes are exposed in the API (which is a surprisingly common thing). Again, japicmp is incapable of doing that.

That is not to say that I don't want to make it quicker...

**Comment by rsvoboda** [comment]
Command is still running with Java 9, started 40+ minutes ago :(
I think I see https://github.com/revapi/revapi/issues/112 here

CPU is not stressed as much as with Java 8 where all 8 cores were used.
Now it seems just 4 are used and only for 50%.

**Comment by rsvoboda** [comment]
Java 9.0.4:
```
java.lang.IllegalStateException: Failed to obtain class tree due to compilation failure.
	at org.revapi:revapi-java:jar:0.16.0//org.revapi.java.model.JavaElementForest.waitForCompilation(JavaElementForest.java:125)
	at org.revapi:revapi-java:jar:0.16.0//org.revapi.java.model.JavaElementForest.getRoots(JavaElementForest.java:59)
	at org.revapi.Revapi.analyzeWith(Revapi.java:258)
	at org.revapi.Revapi.analyze(Revapi.java:184)
	at org.revapi.standalone.Main.run(Main.java:378)
	at org.revapi.standalone.Main.main(Main.java:257)
Caused by: java.lang.RuntimeException: java.lang.OutOfMemoryError: Java heap space
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:158)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
	at jdk.compiler/com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:71)
	at org.revapi:revapi-java:jar:0.16.0//org.revapi.java.compilation.ProbingAnnotationProcessor$1.call(ProbingAnnotationProcessor.java:80)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.OutOfMemoryError: Java heap space

real	104m51.186s
user	186m28.010s
sys	1m25.902s
```

**Comment by metlos** [comment]
Note that #112 is not the sole cause of the slowness you're seeing. While I identified a couple of areas I really *should* improve (and already addressed the one mentioned in #112), the java compiler API simply got much slower (I assume that is due to the module system in 9 and onward).

The only solution for that would be to rewrite Revapi to not use the java compiler which is a humongous task whichever way I slice it.

So in another words, I will do what I can in Revapi itself to speed things up, but it will always be slower on Java 9+ due to its use of java compiler to class analysis.

**Comment by metlos** [comment]
I'm closing this because there is no other way forward but to move away from the internal use of java compiler to do the classfile reading and declaration inspections.
